feat: support `as` annotations for init parameters

i582 opened the pull request:
## Issue

Closes #1023.

## Checklist

- [x] I have updated CHANGELOG.md
- [x] I have added tests to demonstrate the contribution is correctly implemented: this usually includes both positive and negative tests, showing the happy path(s) and featuring intentionally broken cases

i582 committed:
feat: support `as` annotations for init parameters

Fixes #1023

i582 committed:
Merge branch 'main' into pmakhnev/init-as-params

i582 committed:
fix

i582 committed:
changelog

i582 committed:
add remaining types handling

i582 committed:
add map to codegen test

anton-trunov reviewed (changes requested):
- Formatter and AST printer should support the new syntax
- Negative tests like: `init(x: Int as uint260) { ... }`
- Add an e2e test like this one `src/test/e2e-emulated/contract-init/contract-with-init-init-parameter.tact`
- Docs
- Use in `src/benchmarks/contracts/sbt/sbt-item.tact`
- Use in `src/benchmarks/contracts/nft-item.tact`

i582 committed:
tests: formatter, e2e

i582 committed:
test for invalid as type

i582 commented:
> * Formatter and AST printer should support the new syntax

Already support, added tests

> * Negative tests like: `init(x: Int as uint260) { ... }`

Added

> * Add an e2e test like this one `src/test/e2e-emulated/contract-init/contract-with-init-init-parameter.tact`

Added

> * Use in `src/benchmarks/contracts/sbt/sbt-item.tact`
> * Use in `src/benchmarks/contracts/nft-item.tact`

We don't use `init` here anymore, so there's no need to change anything :)

i582 committed:
Merge branch 'main' into pmakhnev/init-as-params

i582 committed:
fmt

i582 committed:
doc

anton-trunov commented:
what about `init()` in `src/benchmarks/contracts/nft-collection.tact`? ;)

i582 committed:
add as for NFTCollection init

i582 commented:
> what about `init()` in `src/benchmarks/contracts/nft-collection.tact`? ;)

Ooops, added :)

i582 committed:
fix
